File: src/main/java/com/kishlaly/ta/analyze/tasks/blocks/one/Long_ScreenOne_MACD_Last_X_Ascending.java
package com.kishlaly.ta.analyze.tasks.blocks.one;

import com.kishlaly.ta.analyze.tasks.ThreeDisplays;
import com.kishlaly.ta.model.BlockResultJava;
import com.kishlaly.ta.model.SymbolDataJava;
import com.kishlaly.ta.model.indicators.IndicatorJava;
import com.kishlaly.ta.model.indicators.MACDJava;
import com.kishlaly.ta.utils.LogJava;

import java.util.List;

import static com.kishlaly.ta.model.BlockResultCodeJava.OK;
import static com.kishlaly.ta.model.BlockResultCodeJava.X_HISTOGRAMS_NOT_ASCENDING_SCREEN_1;

/**
 * the last X histograms grow consecutively
 */
public class Long_ScreenOne_MACD_Last_X_Ascending implements ScreenOneBlockJava {
    @Override
    public BlockResultJava check(SymbolDataJava screen) {
        if (ThreeDisplays.Config.EMA26_TOTAL_BARS_CHECK < 0) {
            throw new RuntimeException("ThreeDisplays.Config.EMA26_TOTAL_BARS_CHECK not set");
        }
        List<MACDJava> screen_1_MACD = (List<MACDJava>) screen.indicators.get(IndicatorJava.MACD);
        int count = 0;
        for (int i = screen_1_MACD.size() - ThreeDisplays.Config.EMA26_TOTAL_BARS_CHECK; i < screen_1_MACD.size() - 1; i++) {
            Double currentValue = screen_1_MACD.get(i).getHistogram();
            Double nextValue = screen_1_MACD.get(i + 1).getHistogram();
            if (currentValue < nextValue) {
                count++;
            }
        }

        if (count < ThreeDisplays.Config.EMA26_TOTAL_BARS_CHECK) {
            LogJava.recordCode(X_HISTOGRAMS_NOT_ASCENDING_SCREEN_1, screen);
            LogJava.addDebugLine(ThreeDisplays.Config.EMA26_TOTAL_BARS_CHECK + " histograms do not grow on the long-term screen");
            return new BlockResultJava(screen.getLastQuote(), X_HISTOGRAMS_NOT_ASCENDING_SCREEN_1);
        }
        return new BlockResultJava(screen.getLastQuote(), OK);
    }

}

File: src/main/kotlin/com/kishlaly/ta/analyze/tasks/blocks/one/Long_ScreenOne_MACD_Last_X_Ascending.kt
package com.kishlaly.ta.analyze.tasks.blocks.one

import com.kishlaly.ta.analyze.tasks.ThreeDisplays
import com.kishlaly.ta.model.BlockResult
import com.kishlaly.ta.model.BlockResultCode
import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.model.indicators.Indicator
import com.kishlaly.ta.model.indicators.MACD
import com.kishlaly.ta.utils.Log

/**
 * the last X histograms grow consecutively
 */
class Long_ScreenOne_MACD_Last_X_Ascending : ScreenOneBlock {
    override fun check(screen: SymbolData): BlockResult {
        if (ThreeDisplays.Config.EMA26_TOTAL_BARS_CHECK < 0) {
            println("ThreeDisplays.Config.EMA26_TOTAL_BARS_CHECK not set")
            return BlockResult.configurationError(screen.lastQuote)
        }
        val screen_1_MACD = screen.indicator(Indicator.MACD) as List<MACD>
        var count = 0
        for (i in screen_1_MACD.size - ThreeDisplays.Config.EMA26_TOTAL_BARS_CHECK until screen_1_MACD.size - 1) {
            val currentValue = screen_1_MACD[i].histogram
            val nextValue = screen_1_MACD[i + 1].histogram
            if (currentValue < nextValue) {
                count++
            }
        }
        if (count < ThreeDisplays.Config.EMA26_TOTAL_BARS_CHECK) {
            Log.recordCode(BlockResultCode.X_HISTOGRAMS_NOT_ASCENDING_SCREEN_1, screen)
            Log.addDebugLine(ThreeDisplays.Config.EMA26_TOTAL_BARS_CHECK.toString() + " histograms do not grow on the long-term screen")
            return BlockResult(screen.lastQuote, BlockResultCode.X_HISTOGRAMS_NOT_ASCENDING_SCREEN_1)
        }
        return BlockResult(screen.lastQuote, BlockResultCode.OK)
    }
}